Intermine gene list import/export API for scalatra

jtnystrom opened the pull request:
This PR adds an API for exporting/importing named gene sets and the ability to configure an Intermine instance using environment variables. TargetMine connection details were added to .envrc.example.

To use these calls it is also necessary to supply username and password for the Intermine instance using the 'user' and 'pass' parameters. An account for TargetMine may be created on the TargetMine website https://targetmine.mizuguchilab.org.

Example data for import/export:

```json 
[
{
    "name": "glutathione",
    "items": [
      "1367834_at",
      "XM_003749458",      
      "XM_017591333",
      "1368354_at",
      "XM_008772860",
      "NM_053293", ...
  ] },
 {
    "name": "asdf2",
    "items": [
      "1433133_at",
      "1440162_x_at",
      "1433132_at",
      "1437800_at", ...
 ] }
]
```

Prior to export, gene lists are converted from probes (as in the example above) to entrez genes (they are stored in this format on the intermine instance), and when we import they are converted back into probes.
On my system this means that lists may grow considerably when imported back, as the entrez gene <-> probe mapping is many to many, and I have many different platforms  (any match in any platform would be included in the result). If this is becomes a problem, the effect might be avoided by supplying a desired platform or a representative sample group to the import call, to restrict the reverse conversion to relevant platforms only.

Example import request:
 GET http://localhost:4200/json/intermine/list?user=user@email.com&pass=xyz&platform=Rat230_2 - returns json data like the above (assuming Rat230_2 is a platform in the database)

Example export request:
POST http://localhost:4200/json/intermine/list?user=user@email.com&pass=xyz&replace=true with json data (example above) in the post body. All parameters are mandatory.

The following was tested:
* Importing gene lists (curl)
* Exporting gene lists (curl)
* Ensuring that exports with the replace=false flag does not overwrite existing lists with the same name
* Importing and exporting gene lists from the GWT application, ensuring that old functionality was not broken by these changes
* Importing with different target platforms, verifying that the results were correct

jtnystrom committed:
Carry out intermine list import/export from IntermineConnector without using the deprecated StringList type.
Simplify some intermine related functions, removing unused parameters.

jtnystrom committed:
Add GeneList class.
Add IntermineHandling for handling intermine requests from scalatra. Read intermine connection settings from environment variables.
Skeleton implementations of get "intermine/list" and post "/intermine/list" for import and export.

jtnystrom committed:
Move GeneList to t.server.viewer.intermine and consistently use it to represent named gene lists in method signatures.

jtnystrom committed:
Add TargetMine connection details to .envrc.example

yujiberra commented:
@jtnystrom 
Thanks for your work on this! 
Could you also write out the specification of the API in the PR body?

yujiberra commented:
I see your edits, thanks.

Also, confirming just in case but the example data is two array items and you omitted the opening bracket of the array right?

jtnystrom commented:
> Also, confirming just in case but the example data is two array items and you omitted the opening bracket of the array right?
Correct.

jtnystrom commented on Toxygates/src/t/gwt/viewer/client/intermine/InterMineData.java:
```diff
@@ -62,7 +62,7 @@ public InterMineData(ImportingScreen parent, @Nullable IntermineInstance preferr
   /**
    * PreferredInstance must not be null
    */
-  public void importLists(final boolean asProbes) {
```
Removed this parameter in a number of places since it was always true, and no logic for the false case had ever been implemented.

RodolfoAllendes commented on Toxygates/src/t/server/viewer/intermine/Intermine.scala:
```diff
@@ -67,71 +67,100 @@ class IntermineConnector(instance: IntermineInstance,
+
+  def exportLists(user: String, pass: String, lists: Iterable[GeneList], replace: Boolean): Unit =
+    try {
+      val ls = getListService(Some(user), Some(pass))
```
I have only one issue with providing the user's password... AFAIK to avoid this, intermine provides the use of access tokens, that can be used as an alternative to the password, as they are time-limited (although also potentially permanent) ways to access the lists.

For example, with the python API for Intermine, rather than initializing a service using user/pass as shown below:
`service = Service(service, username=user, password=pwd)`
you would do it with
`service = Service(service, token=token)`

After defining the service, you actually access/create list using appropriate methods on the defined service.. something along the lines of 
`service.create_list(genes, name=lname, list_type='Gene')`
(which is what you are doing here in line 131... the initialization of the service I didnt see it/find it :sweat_smile:  

So I was wondering if (it wasnt too much work) it would be possible to change the access method

PS: first time using this comments.. so not sure how its going to show up in the review of the PR XD

jtnystrom replied:
Hi Rodolfo, thanks for your comment. I have thought a little about this before and internally, we do use tokens for some of the communication once the session has been established. 
If we use tokens throughout and never send login details, as you suggest, the end user would be responsible for creating and supplying a token, right? Do you think that would be easy for average users to do? My impression is that this would be a more complicated experience than supplying username/password, but I might be wrong.

yujiberra reviewed (commented):
@jtnystrom 
Thanks for your work here! 

First, some questions about InterMine/TargetMine: I understand that TargetMine is one of many InterMines; is it true that if we wanted to support something other than TargetMine, it would just be a matter of changing `T_INTERMINE_URL`? 

I'm just trying to understand the reasoning for using environment variables to set information about the InterMine instance. Do we expect situations where we want to change the InterMine instance without a rebuild? If we supported multiple InterMine instances I could see it being nice to be able to add/remove them without a rebuild, but that would require a somewhat different implementation. 

Also regarding the following:
> Prior to export, gene lists are converted from probes (as in the example above) to entrez genes (they are stored in this format on the intermine instance), and when we import they are converted back into probes.
On my system this means that lists may grow considerably when imported back, as the entrez gene <-> probe mapping is many to many, and I have many different platforms (any match in any platform would be included in the result). If this is becomes a problem, the effect might be avoided by supplying a desired platform or a representative sample group to the import call, to restrict the reverse conversion to relevant platforms only.

as you know, in the front end the platform (and omics type) is part of the gene set's definition. It's not clear what the front end should do with a gene set imported from intermine, since it'll include probes from many platform.

I think the easiest solution would be to specify a platform on the front end when importing gene sets from intermine, which would require the following changes/additions to the back end:

- add a `platform` parameter to GET `intermine/list`, and restrict its output to genes from the platform
- I think I also need a way to get a list of platforms along with their omics type; I realize that there's a way to list platforms but it doesn't provide omics type

What do you think?

jtnystrom commented:
> I'm just trying to understand the reasoning for using environment variables to set information about the InterMine instance. Do we expect situations where we want to change the InterMine instance without a rebuild? If we supported multiple InterMine instances I could see it being nice to be able to add/remove them without a rebuild, but that would require a somewhat different implementation.

It's correct that changing the InterMine connection is only a matter of changing the environment variables here.
In the GWT application, we actually support four different instances: humanmine, ratmine, targetmine, mousemine.
They are configured in web.xml, and any number can be added.
For the Scalatra application, I'm not sure what we will ultimately want to do yet, but I imagine that we might want to have a similar system with multiple supported instances in the future, although it's not something we need urgently. Hence configuring by environment variables as a kind of groundwork for that. Though it would not be very different to just hardcode the connection.

jtnystrom commented:
> I think the easiest solution would be to specify a platform on the front end when importing gene sets from intermine, which would require the following changes/additions to the back end:
> 
>     * add a `platform` parameter to GET `intermine/list`, and restrict its output to genes from the platform
> 
>     * I think I also need a way to get a list of platforms along with their omics type; I realize that there's a way to list platforms but it doesn't provide omics type

If you have sample group definitions, you probably also have sample objects that you have obtained from the server, and they are tagged with platforms and omics types. We always assume that a group has a single omics type and platform. So I think the easiest thing to do extract a platform from a sample in a group that the user has already defined, and then pass that as a platform parameter (as you suggested) when we do the list import? (And then the user can be told that the import may have a different result depending on which sample group you import for)

jtnystrom commented:
Discussed with @yujiberra and agreed to create a /platform endpoint to obtain a list of platforms, and to supply this as a required parameter when doing the list import.

jtnystrom committed:
Add a /platform/user call and share implementation with /platform, to get a list of user-facing platforms.
Add a "type" field to the platform list.

jtnystrom committed:
Require that the user supplies a preferred platform when importing InterMine gene lists with /intermine/list.
This avoids the problem where the size of the imported list can explode due to many platforms in the system. It is also much faster.

jtnystrom commented on OTGTool/src/main/scala/t/platform/PlatformRegistry.scala:
```diff
@@ -97,17 +110,17 @@ class PlatformRegistry(loader: PlatformLoader) {
-    var rem = Set() ++ probes
-    var r = Set[String]()
-    for (p <- platforms; valid = filterProbes(rem, p, species)) {
-      rem --= valid
```
This logic with a "remaining" (rem) set to be filtered, which is gradually subtracted from, is a legacy concept that is no longer needed here. Probes belong to exactly one platform, and I don't think there was much efficiency to be gained by complicating the method like this

jtnystrom commented on OTGTool/src/main/scala/t/sparql/PlatformStore.scala:
```diff
@@ -87,12 +87,12 @@ class PlatformStore(val config: TriplestoreConfig) extends ListManager(config) {
-  def platformTypes: Map[String, String] = {
-    Map() ++ triplestore.mapQuery(s"$tPrefixes\nSELECT ?l ?type WHERE { ?item a $itemClass; rdfs:label ?l ; " +
-      s"$platformType ?type } ").map(x => {
+  def platformOmicsTypes: Map[String, String] = {
```
The old version of this method "platformTypes" wasn't used any longer.

yujiberra reviewed (approved):
@jtnystrom 
> For the Scalatra application, I'm not sure what we will ultimately want to do yet, but I imagine that we might want to have a similar system with multiple supported instances in the future, although it's not something we need urgently. Hence configuring by environment variables as a kind of groundwork for that. Though it would not be very different to just hardcode the connection.

Yeah that makes sense, thanks for the explanation. I personally wouldn't have done this now because
- it now means we have to set an environment variable that's probably always going to be the same value
- when we do want to support multiple Intermine instances we'll have to use different environment variables (or at least interpret the existing one in a different way)

but it's not such a big deal that I think it should be changed before merging this PR or anything.

In any case, I tested this PR locally and was able to do the following, though I didn't check the resulting gene sets in detail:
- retrieve gene sets from TargetMine by specifying Rat230_2 as platform
- upload a gene set to TargetMine
- get a list of platforms from `/platform/user`

I also only took a quick look at the code but things looked fine.

LGTM!

jtnystrom commented:
@yujiberra Thanks for your review and for testing. I agree about the intermine instances, let's consider that a provisional thing. I might hardcode it as part of some future PR.
